src/raknetserver/pipeline/raknet/RakNetPacketReliabilityHandler.java

```diff
@@ -16,59 +18,85 @@
+
 public class RakNetPacketReliabilityHandler extends ChannelDuplexHandler {
 
+	protected static final int WINDOW = 4096;
+	protected static final int HALF_WINDOW = WINDOW / 2;
+	protected static final int RTT_FLOOR = 5; //millis
+
 	protected static final PacketHandlerRegistry<RakNetPacketReliabilityHandler, RakNetPacket> registry = new PacketHandlerRegistry<>();
 	static {
 		registry.register(RakNetEncapsulatedData.class, (ctx, handler, packet) -> handler.handleEncapsulatedData(ctx, packet));
 		registry.register(RakNetACK.class, (ctx, handler, packet) -> handler.handleAck(ctx, packet));
 		registry.register(RakNetNACK.class, (ctx, handler, packet) -> handler.handleNack(ctx, packet));
 	}
 
+	protected final Channel channel;
+	protected final Int2ObjectOpenHashMap<RakNetEncapsulatedData> sentPackets = new Int2ObjectOpenHashMap<>();
+
+	protected int lastReceivedSeqId = 0;
+	protected int nextSendSeqId = 0;
+	protected long minRTT = 2000;
+
+	public RakNetPacketReliabilityHandler(Channel channel) {
+		this.channel = channel;
+		startResendTimer();
+	}
+
+	private void startResendTimer() {
+		channel.eventLoop().schedule(() -> {
+			if (channel.isOpen()) {
+				startResendTimer();
+				resendTick();
+			}
+		}, minRTT, TimeUnit.MILLISECONDS);
+	}
+
 	@Override
-	public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+	public void channelRead(ChannelHandlerContext ctx, Object msg) {
 		if (msg instanceof RakNetPacket) {
 			registry.handle(ctx, this, (RakNetPacket) msg);
 		} else {
 			ctx.fireChannelRead(msg);
 		}
 	}
 
-	protected static final int HALF_WINDOW = UINT.B3.MAX_VALUE / 2;
-
-	protected final Int2ObjectOpenHashMap<RakNetEncapsulatedData> sentPackets = new Int2ObjectOpenHashMap	<>();
-	protected int lastReceivedSeqId = -1;
+	protected boolean idWithinWindow(int id) {
+		return Math.abs(UINT.B3.minusWrap(id, lastReceivedSeqId)) < HALF_WINDOW;
+	}
 
 	protected void handleEncapsulatedData(ChannelHandlerContext ctx, RakNetEncapsulatedData packet) {
-		RakNetEncapsulatedData edata = packet;
-		int packetSeqId = edata.getSeqId();
-		int seqIdDiff = UINT.B3.minus(packetSeqId, lastReceivedSeqId);
-		//ignore duplicate packet
-		if ((seqIdDiff == 0) || (seqIdDiff > HALF_WINDOW)) {
+		int packetSeqId = packet.getSeqId();
+		ctx.writeAndFlush(new RakNetACK(packetSeqId)).addListener(ChannelFutureListener.FIRE_EXCEPTION_ON_FAILURE);
+		if (!idWithinWindow(packetSeqId)) { //ignore duplicate packet
 			return;
 		}
-		//send nack for missed packets
-		if (seqIdDiff > 1) {
-			ctx.writeAndFlush(new RakNetNACK(UINT.B3.plus(lastReceivedSeqId, 1), UINT.B3.minus(packetSeqId, 1))).addListener(ChannelFutureListener.FIRE_EXCEPTION_ON_FAILURE);
+		if (UINT.B3.minusWrap(packetSeqId, lastReceivedSeqId) > 0) { //can be zero on the first packet only
+			lastReceivedSeqId = UINT.B3.plus(lastReceivedSeqId, 1);
+			while (lastReceivedSeqId != packetSeqId) { //nack any missed packets before this one
```

Comment:
For loop for literally no point

Reply:
still need to send a nack for packet id

Reply:
Or maybe send nack for range prev + 1 <-> last -1 like it was before. it doesn't require a for loop since ack/nack accept ranges.

Reply:
true